newIDE/app/src/MainFrame/Toolbar/PreviewAndPublishButtons.js

```diff
@@ -0,0 +1,202 @@
+  return (
+    <ThemeConsumer>
+      {muiTheme => (
+        <React.Fragment>
+          <ElementWithMenu
+            element={
+              <ToolbarIcon
+                disabled={!isPreviewEnabled}
+                src="res/ribbon_default/bug32.png"
+                tooltip={t`Advanced preview options (debugger, network preview...)`}
+              />
+            }
+            buildMenuTemplate={(i18n: I18nType) => [
+              {
+                label: i18n._(t`Start Network Preview (Preview over WiFi/LAN)`),
+                click: onNetworkPreview,
+                enabled: canDoNetworkPreview,
+              },
+              { type: 'separator' },
+              {
+                label: i18n._(
+                  t`Start Preview with Debugger and Performance Profiler`
+                ),
+                click: onOpenDebugger,
+              },
+            ]}
+          />
+          <ElementWithMenu
+            element={
+              <FlatButton
+                onClick={onHotReloadPreview}
+                disabled={!isPreviewEnabled}
+                icon={
+                  <img
+                    alt="Preview"
+                    src={
+                      hasPreviewsRunning
+                        ? 'res/ribbon_default/hotReload64.png'
+                        : previewState.isPreviewOverriden
+                        ? 'res/ribbon_default/previewOverride32.png'
+                        : 'res/ribbon_default/preview64.png'
+                    }
+                    width={32}
+                    height={32}
+                    style={{
+                      filter: !isPreviewEnabled
+                        ? 'grayscale(100%)'
+                        : muiTheme.gdevelopIconsCSSFilter,
+                    }}
+                  />
+                }
+                label={
+                  hasPreviewsRunning ? (
+                    <Trans>Update</Trans>
+                  ) : (
+                    <Trans>Preview</Trans>
+                  )
+                }
+                exceptionalTooltipForToolbar={
+                  hasPreviewsRunning ? (
+                    <Trans>
+                      Apply changes to the running preview, right click for more
+                    </Trans>
+                  ) : previewState.isPreviewOverriden ? (
+                    <Trans>Preview is overridden, right click for more</Trans>
+                  ) : previewState.previewExternalLayoutName ? (
+                    <Trans>
+                      Launch a preview of the external layout inside the scene,
+                      right click for more
+                    </Trans>
+                  ) : (
+                    <Trans>
+                      Launch a preview of the scene, right click for more
+                    </Trans>
+                  )
+                }
+              />
+            }
+            openMenuWithSecondaryClick
+            buildMenuTemplate={(i18n: I18nType) => [
```

Comment:
What about extracting these 2 buildMenuTemplate into callbacks to improve the component's readability ?